File: baekjoon/1753.py
import sys
from queue import PriorityQueue

Edge = tuple[
    int,  # Next node
    int,  # Weight
]

Node = list[Edge]

Cost = tuple[
    int,  # Cost
    int,  # Destination node
]

INFINITE = 1_000_000


def get_costs(nodes: list[Node], start_node: int) -> list[int]:
    costs = [INFINITE for _ in nodes]
    p_queue = PriorityQueue[Cost]()

    costs[start_node] = 0
    p_queue.put((0, start_node))

    while not p_queue.empty():
        current_cost, current_node = p_queue.get()
        if costs[current_node] < current_cost:
            continue
        for next_node, weight in nodes[current_node]:
            old_cost = costs[next_node]
            new_cost = current_cost + weight
            if new_cost < old_cost:
                costs[next_node] = new_cost
                p_queue.put((new_cost, next_node))

    return costs


def main():
    node_count, edge_count = (int(s) for s in input().split())
    start_node = int(input()) - 1
    nodes: list[Node] = [[] for _ in range(node_count)]
    for _ in range(edge_count):
        from_node, to_node, weight = (int(s) for s in sys.stdin.readline().split())
        from_node -= 1
        to_node -= 1
        edge: Edge = (to_node, weight)
        nodes[from_node].append(edge)
    costs = get_costs(nodes, start_node)
    for cost in costs:
        sys.stdout.write((str(cost) if cost != INFINITE else "INF") + "\n")


main()

File: baekjoon/1753.cpp
#include <iostream>
#include <queue>

using namespace std;

struct Edge {
  int next;
  int weight;
};

struct Node {
  vector<Edge> edges;
};

struct Cursor {
  int cost;
  int destination;
};

const int INF = 1000000007;

auto operator<(const Cursor& obj_a, const Cursor& obj_b) -> bool {
  if (obj_a.cost != obj_b.cost) {
    return obj_a.cost < obj_b.cost;
  }
  return obj_a.destination < obj_b.destination;
}

auto operator>(const Cursor& obj_a, const Cursor& obj_b) -> bool {
  if (obj_a.cost != obj_b.cost) {
    return obj_a.cost > obj_b.cost;
  }
  return obj_a.destination > obj_b.destination;
}

auto get_costs(vector<Node>& nodes, int start_node) -> vector<int> {
  int node_count = static_cast<int>(nodes.size());
  vector<int> costs(node_count, INF);
  priority_queue<Cursor, vector<Cursor>, greater<>> p_queue;

  costs[start_node] = 0;
  p_queue.push({0, start_node});

  while (!p_queue.empty()) {
    Cursor cursor = p_queue.top();
    p_queue.pop();
    for (Edge edge : nodes[cursor.destination].edges) {
      int written_cost = costs[edge.next];
      int new_cost = cursor.cost + edge.weight;
      if (new_cost < written_cost) {
        costs[edge.next] = new_cost;
        p_queue.push({new_cost, edge.next});
      }
    }
  }

  return costs;
}

auto main() -> int {
  int node_count = 0;
  int edge_count = 0;
  cin >> node_count >> edge_count;
  int start_node = 0;
  cin >> start_node;
  start_node -= 1;

  vector<Node> nodes(node_count, Node());
  for (int i = 0; i < edge_count; i++) {
    int from_node_raw = 0;
    int to_node_raw = 0;
    int weight = 0;
    cin >> from_node_raw >> to_node_raw >> weight;
    nodes[from_node_raw - 1].edges.push_back({to_node_raw - 1, weight});
  }

  vector<int> costs = get_costs(nodes, start_node);
  for (int cost : costs) {
    if (cost < INF) {
      cout << cost << '\n';
    } else {
      cout << "INF\n";
    }
  }
}